Better betterness

BillWagner opened the pull request:
Fixes #157
Fixes #283 

This takes the text from [Better betterness](https://github.com/dotnet/roslyn/blob/main/docs/specs/CSharp%206/Better%20Betterness.md) and then updates it for compatibility fixes, related to exact matches and method group conversions.

Reviewing each commit in order clarifies what changes came from the roslyn doc, and then those changes that followed after incorporating that initial spec. Commit messages link to dotnet/roslyn#6750 that explains the later changes.

BillWagner committed:
Incorporate Better Betterness

This commit adds the notes from https://github.com/dotnet/roslyn/blob/main/docs/specs/CSharp%206/Better%20Betterness.md

BillWagner committed:
Incorporate changes from #283

Issue #283 contains notes for how the implementation differed from the proposed better betterness text.

BillWagner committed:
compat for delegate conversion

This commit brings in the method group conversion fix that caused dotnet/roslyn#6750

It needs wordsmithing for a number of reasons:
- "corresponding argument" isn't defined.
- the section references itself.
- I think there may have been requirements from the prior commit regarding delegates and expressions that are still needed.

BillWagner commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
-  - `T₁` is `int` and `T₂` is `uint`, or `ulong`
-  - `T₁` is `long` and `T₂` is `ulong`
+- An implicit conversion from `T₁` to `T₂` exists
+- In case of a method group conversion (§10.6) for the corresponding argument, if a better conversion target (§7.5.3.5), is a delegate type that is not compatible (§19.4) with the single best method from the method group (§10.6), then neither delegate type is better.
```
This bullet item still needs work (where it may be more than one bullet item)

jskeet replied:
Right - "In case of a method group conversion" definitely seems like an odd start to this bullet.

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
-Given an implicit conversion `C₁` that converts from an expression `E` to a type `T₁`, and an implicit conversion `C₂` that converts from an expression `E` to a type `T₂`, `C₁` is a ***better conversion*** than `C₂` if at least one of the following holds:
+Given an implicit conversion `C₁` that converts from an expression `E` to a type `T₁`, and an implicit conversion `C₂` that converts from an expression `E` to a type `T₂`, `C₁` is a better conversion than `C₂` if one of the following holds:
+- `E` exactly matches `T₁` and `E` does not exactly match `T₂` (§11.6.4.5)
+- `E` exactly matches both or neither of `T₁` and `T2`, and `T₁` is a better conversion target than `T2` (§11.6.4.6)
```
Nit: T2 should be T₂.

More importantly, this looks different to the Roslyn doc, in terms of the second bullet and E exactly matching T₂.

Is this effectively a fix to what was written in the Roslyn doc?

BillWagner replied:
> More importantly, this looks different to the Roslyn doc, in terms of the second bullet and E exactly matching T₂.
> 
> Is this effectively a fix to what was written in the Roslyn doc?

I think so. I added a note in the PR description relating to this. The first commit brings in the roslyn doc. Later commits address issues with that spec language.

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
-    - `D₁` has a return type `Y₁`, and `D₂` has a return type `Y₂`, an inferred return type `X` exists for `E` in the context of that parameter list ([§11.6.3.13](expressions.md#116313-inferred-return-type)), and the conversion from `X` to `Y₁` is better than the conversion from `X` to `Y₂`
-    - `E` is async, `D₁` has a return type `Task<Y₁>`, and `D₂` has a return type `Task<Y>`, an inferred return type `Task<X>` exists for `E` in the context of that parameter list ([§11.6.3.13](expressions.md#116313-inferred-return-type)), and the conversion from `X` to `Y₁` is better than the conversion from `X` to `Y₂`
-    - `D₁` has a return type `Y`, and `D₂` is void returning
+### 11.6.4.5 Exactly matching Expression
```
Is there a reason for Expression to have a capital E here (it doesn't in the preceding section)

jskeet replied:
Should this be `####` rather than `###`?

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+- `E` has a type `S`, and an identity conversion exists from `S` to `T`
+- `E` is an anonymous function, `T` is either a delegate type `D` or an expression tree type `Expression<D>` and one of the following holds:
+  - An inferred return type `X` exists for `E` in the context of the parameter list of `D` (§11.6.3.12), and an identity conversion exists from `X` to the return type of `D`
+  - Either `E` is non-async and `D` has a return type `Y` or `E` is async and  `D` has a return type `Task<Y>`, and one of the following holds:
```
How does that work for `ValueTask<Y>` etc?

BillWagner replied:
I think it will work the same. (Generalized async return types don't come until V7 though).

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+  - An inferred return type `X` exists for `E` in the context of the parameter list of `D` (§11.6.3.12), and an identity conversion exists from `X` to the return type of `D`
+  - Either `E` is non-async and `D` has a return type `Y` or `E` is async and  `D` has a return type `Task<Y>`, and one of the following holds:
+    - The body of `E` is an expression that exactly matches `Y`
+    - The body of `E` is a statement block where every return statement returns an expression that exactly matches `Y`
```
Do we need "statement block" rather than just "block"?

BillWagner replied:
ping @RexJaeschke  to check if "statement block" or "block" is better.  

From Jon: There are two other occurrences of "statement block" that should be changed.  Jon will create an issue for that change.

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+### 11.6.4.5 Exactly matching Expression
 
-#### 11.6.4.5 Better conversion from type
+Given an expression `E` and a type `T`, `E` ***exactly matches*** `T` is one of the following holds:
```
Nit: is => if

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
 
--   An identity conversion exists from `S` to `T₁` but not from `S` to `T₂`
--   `T₁` is a better conversion target than `T₂` ([§11.6.4.6](expressions.md#11646-better-conversion-target))
+### 11.6.4.6 Better conversion target
```
Again, I suspect this should be `####` (or 11.6.4.4 is wrong)

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+### 11.6.4.6 Better conversion target
 
-#### 11.6.4.6 Better conversion target
+Given two different types `T₁` and `T₂`, `T₁` is a ***better conversion target*** than `T₂` if:
```
We need something like "if at least one of the following holds" or "if all of the following hold"

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+- An implicit conversion from `T₁` to `T₂` exists
+- In case of a method group conversion (§10.6) for the corresponding argument, if a better conversion target (§7.5.3.5), is a delegate type that is not compatible (§19.4) with the single best method from the method group (§10.6), then neither delegate type is better.
+- `T₁` is `Task<S₁>`, `T₂` is `Task<S₂>`, and `S₁` is a better conversion target than `S₂`
+- `T₁` is `S₁` or `S₁`? where `S₁` is a signed integral type, and `T₂` is `S₂` or `S₂`? where `S₂` is an unsigned integral type. Specifically:
```
Should the question-mark be within the back-ticks in both occurrences of this?

jskeet reviewed (commented):
Some initial thoughts that didn't require actual understanding. I haven't tried to work out what situations this actually helps with yet :)

BillWagner committed:
initial feedback.

BillWagner committed:
section reference

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
-  - `D₁` and `D₂` have identical parameter lists, and one of the following holds:
-    - `D₁` has a return type `Y₁`, and `D₂` has a return type `Y₂`, an inferred return type `X` exists for `E` in the context of that parameter list ([§11.6.3.13](expressions.md#116313-inferred-return-type)), and the conversion from `X` to `Y₁` is better than the conversion from `X` to `Y₂`
-    - `E` is async, `D₁` has a return type `Task<Y₁>`, and `D₂` has a return type `Task<Y>`, an inferred return type `Task<X>` exists for `E` in the context of that parameter list ([§11.6.3.13](expressions.md#116313-inferred-return-type)), and the conversion from `X` to `Y₁` is better than the conversion from `X` to `Y₂`
-    - `D₁` has a return type `Y`, and `D₂` is void returning
```
In the current Standard we have delegate return type of `Y`, presumably non-void, being a better conversion than a return type of void. I don’t immediately see this case exists in the revised model, is this intentional? A breaking change?

Nigel-Ecma replied:
I've now found the void case is in the source material [Better betterness](https://github.com/dotnet/roslyn/blob/main/docs/specs/CSharp%206/Better%20Betterness.md) so I guess this was unintentional?

BillWagner replied:
This change is intentional, and it was a regression in the original Better Betterness spec. The issue is dotnet/roslyn#6750.  There is still work to do here, because I think this loses some other requirements.

jskeet replied:
Case that I think is broken by this:

```csharp
void M(Action action) {}
void M(Func<object> func) {}

void Test()
{
    // "xyz" doesn't *exactly match* object, so this should be ambiguous (according to the PR)
    M(() => "xyz");
}
```

Case that I think isn't:

```csharp
void M(Action action) {}
void M(Func<object> func) {}

void Test()
{
    // new object() exactly matches object, therefore Func<object> is preferred
    M(() => new object());
}
```

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -945,34 +945,33 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
-  - `T₁` is `int` and `T₂` is `uint`, or `ulong`
-  - `T₁` is `long` and `T₂` is `ulong`
+- An implicit conversion from `T₁` to `T₂` exists
+- In case of a method group conversion (§10.6) for the corresponding argument, if a better conversion target (§11.6.4.6), is a delegate type that is not compatible (§19.4) with the single best method from the method group (§10.6), then neither delegate type is better.
```
I think the comma in “conversion target (§11.6.4.6), is a delegate type” should not be there? And maybe the one in “method group (§10.6), then neither delegate” as well?

BillWagner committed:
a bit of editing

BillWagner committed:
I think this gets close

This gets close to what's implemented in C# 6.0.

I think this fails as a description in the case of #499

For 7.3, it would be better to describe that change in the section on Method Group conversion.

BillWagner commented:
Adding a comment on the roadmap for much of what I've been reading to reach this point.

- [The better betterness spec](https://github.com/dotnet/roslyn/pull/7934) for C# 6.0. This includes the text for the compatibility fix, and an explanation of the new sections. This PR did not include any of the informative text from that document. We may want to consider some of that.
- [Breaking change due to better betterness](https://github.com/dotnet/roslyn/issues/6560). This describes the case for the later changes to the better betterness spec.
- [Implementation of the C# 6.0 rules](https://github.com/dotnet/roslyn/pull/7727).  This PR contains the changes for the compatibility fix. "Adjust Better Betterness rules around method group conversions in order to restore backward compatibility"
- [Additional change for C# 7.3](https://github.com/dotnet/roslyn/issues/8942) This issue describes another case regarding method group conversions. This was fixed in C# 7.3, as described in the issue. I created #499 to track that when we're working on the C# 7.x spec.

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,34 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+  - `S₁` is `short` and `S₂` is `ushort`, `uint`, or `ulong`
+  - `S₁` is `int` and `S₂` is `uint`, or `ulong`
+  - `S₁` is `long` and `S₂` is `ulong`
+- `E` is the result of a method group conversion is a delegate creation expression (§11.7.15.6) whose argument is a method group and `T₁` is compatible (§19.4) with the single best method from the method group (§10.8)
```
Add "or" after "method group conversion"?

BillWagner replied:
I reworked the sentence a bit.

BillWagner committed:
review from the meeting.

BillWagner commented:
Marking as ready for review, and adding Mads and Neal as reviewers.

Consensus from today's meeting was that this is close, and any future updates can be handled in a separate PR (for C# 6)

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,34 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+  - `S₁` is `short` and `S₂` is `ushort`, `uint`, or `ulong`
+  - `S₁` is `int` and `S₂` is `uint`, or `ulong`
+  - `S₁` is `long` and `S₂` is `ulong`
+- `E` is a delegate creation expression (§11.7.15.6) resulting from the result of a method group conversion and `T₁` is compatible (§19.4) with the single best method from the method group (§10.8)
```
Do we really only want this for delegate creation expressions such as `new Action(Method)` rather than plain method group conversions? I suspect it should actually cover both - or possibly *just* method group conversions, given that if there's a delegate creation expression we know the exact type anyway.

Separately, "resulting from the result" sounds redundant and says the same thing twice ;)

BillWagner replied:
> Do we really only want this for delegate creation expressions such as `new Action(Method)` rather than plain method group conversions? I suspect it should actually cover both - or possibly _just_ method group conversions, given that if there's a delegate creation expression we know the exact type anyway.

It should only be plain method group conversions. The `new Action(Method)` is an exact matching expression for `Action`.

I'll reword the sentence.

BillWagner committed:
fix awkward grammar

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,34 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+  - `S₁` is `short` and `S₂` is `ushort`, `uint`, or `ulong`
+  - `S₁` is `int` and `S₂` is `uint`, or `ulong`
+  - `S₁` is `long` and `S₂` is `ulong`
+- `E` is a delegate creation expression (§11.7.15.6) resulting from a method group conversion and `T₁` is compatible (§19.4) with the single best method from the method group (§10.8)
```
Ping for any updates on this? It would be really good to get this PR done and merged this week, so we can get the Word version out next week.

jskeet commented on standard/expressions.md:
```diff
@@ -945,34 +945,34 @@ In case the parameter type sequences `{P₁, P₂, ..., Pᵥ}` and `{Q₁, Q₂
+  - `S₁` is `short` and `S₂` is `ushort`, `uint`, or `ulong`
+  - `S₁` is `int` and `S₂` is `uint`, or `ulong`
+  - `S₁` is `long` and `S₂` is `ulong`
+- `E` is a delegate creation expression (§11.7.15.6) resulting from a method group conversion and `T₁` is compatible (§19.4) with the single best method from the method group (§10.8)
```
Ping for any updates on this? It would be really good to get this PR done and merged this week, so we can get the Word version out next week.

BillWagner replied:
I'm ready to merge, if you are. 

I've pinged Mads offline for a review as well.

jskeet replied:
But this still has "delegate creation expression" where I thought we *just* wanted it to handle method group conversions...

BillWagner replied:
Gotcha. I misunderstood your earlier comment.

Update coming this morning....

BillWagner replied:
Update made.

BillWagner committed:
limit bullet to method group conversion.

jskeet reviewed (approved):
Great - assuming I've remembered correctly that the final line was all I was nervous about, I'm happy with this :)

BillWagner commented:
I'll merge this on the analysis that it's better than what we had before. This area may still need some edits.
